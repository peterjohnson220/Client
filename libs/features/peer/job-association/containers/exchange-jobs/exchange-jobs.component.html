<h2>Exchange Jobs</h2>
<div class="d-flex">
  <i class="fas fa-search"></i>
  <input type="text" placeholder="Accountant" />
  <label>
    <input type="checkbox" />
    Associated
  </label>
  <label>
    <input type="checkbox" />
    Not Associated
  </label>
  <label>Job Family:</label>
  <select>
    <option value="1">Airline</option>
    <option value="2">Accounting</option>
    <option value="3">Restaurant</option>
  </select>
</div>
<label>
  peer exchange jobs: {{ totalExchangeJobs }} total
</label>
<kendo-grid [data]="exchangeJobs">
  <kendo-grid-column field="Id" title="Id"></kendo-grid-column>
  <kendo-grid-column field="JobTitle" title="Job Title"></kendo-grid-column>
  <kendo-grid-column field="JobFamily" title="Job Family"></kendo-grid-column>
  <kendo-grid-column field="ExchangeName" title="Exchange Name"></kendo-grid-column>
</kendo-grid>
