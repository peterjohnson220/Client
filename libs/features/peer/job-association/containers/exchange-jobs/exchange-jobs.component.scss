@import "colors";

.view-button {
  color: $pf-brand-secondary-color;
  border: 0;
  cursor: pointer;
  background-color: #fff;
  opacity: .6;
  transition: opacity .2s ease-in-out;
  font-size: 16px;
  &:hover {
    opacity: 1;
  }
}
.view-button-active {
  color: $pf-brand-secondary-color;
  opacity: 1;
}
