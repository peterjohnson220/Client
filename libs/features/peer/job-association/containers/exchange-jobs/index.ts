export * from './exchange-jobs.component';
