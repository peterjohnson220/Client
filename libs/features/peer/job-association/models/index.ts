export * from './company-job.model';
export * from './exchange-job.model';
