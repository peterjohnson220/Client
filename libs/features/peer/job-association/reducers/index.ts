import { createSelector, createFeatureSelector } from '@ngrx/store';

import { IFeatureGridState } from 'libs/core/reducers/grid.reducer';
import * as fromRoot from 'libs/state/state';
import * as fromGridReducer from 'libs/core/reducers/grid.reducer';
import { GenericMenuItem } from 'libs/models/common';

// Import feature reducers
import * as fromCompanyJobsReducer from './company-jobs.reducer';
import * as fromExchangeJobsReducer from './exchange-jobs.reducer';
import * as fromJobAssociationModalReducer from './job-association-modal.reducer';

// Feature area state
export interface JobAssociationFeatureState {
  companyJobs: IFeatureGridState<fromCompanyJobsReducer.State>;
  exchangeJobs: IFeatureGridState<fromExchangeJobsReducer.State>;
  jobAssociationModal: fromJobAssociationModalReducer.State;
}

// Extend root state with feature area state
export interface State extends fromRoot.State {
  feature_job_association: JobAssociationFeatureState;
}

// Feature area reducers
export const reducers = {
  companyJobs: fromCompanyJobsReducer.reducer,
  exchangeJobs: fromExchangeJobsReducer.reducer,
  jobAssociationModal: fromJobAssociationModalReducer.reducer
};

// Select Feature Area
export const selectFeatureAreaState =
  createFeatureSelector<JobAssociationFeatureState>('feature_job_association');

// Feature Selectors
export const selectExchangeJobsState = createSelector(
  selectFeatureAreaState,
  (state: JobAssociationFeatureState) => state.exchangeJobs);

export const selectCompanyJobsState = createSelector(
  selectFeatureAreaState,
  (state: JobAssociationFeatureState) => state.companyJobs);

export const selectJobAssociationModalState = createSelector(
  selectFeatureAreaState,
  (state: JobAssociationFeatureState) => state.jobAssociationModal);

// Job Association Modal Selectors

export const getJobAssociationModalSaving = createSelector(
  selectJobAssociationModalState,
  fromJobAssociationModalReducer.getSaving);

export const getJobAssociationModalSavingError = createSelector(
  selectJobAssociationModalState,
  fromJobAssociationModalReducer.getSavingError);

// Exchange Jobs Selectors
export const getExchangeJobsGrid = createSelector(
  selectExchangeJobsState,
  (state: IFeatureGridState<fromExchangeJobsReducer.State>) => state.grid);

export const getExchangeJobsGridState = createSelector(
  getExchangeJobsGrid,
  fromGridReducer.getGridState);

export const getExchangeJobsFeature = createSelector(
  selectExchangeJobsState,
  (state: IFeatureGridState<fromExchangeJobsReducer.State>) => state.feature);

export const getExchangeJobsLoading = createSelector(
  getExchangeJobsFeature,
  (feature) => feature.loading
);

export const getExchangeJobsLoadingError = createSelector(
  getExchangeJobsFeature,
  (feature) => feature.loadingError
);

export const getExchangeJobsSearchTerm = createSelector(
  getExchangeJobsFeature,
  (feature) => feature.searchTerm
);

export const getExchangeJobsSelectedExchangeJob = createSelector(
  getExchangeJobsFeature,
  (feature) => feature.selectedExchangeJob
);

export const getExchangeJobsIsDetailPanelExpanded = createSelector(
  getExchangeJobsFeature,
  (feature) => feature.isDetailPanelExpanded
);

export const {
  selectAll: getExchangeJobsList
} = fromExchangeJobsReducer.adapter.getSelectors(getExchangeJobsFeature);

export const getExchangeJobsTotal = createSelector(
  getExchangeJobsFeature,
  fromExchangeJobsReducer.getTotal
);

export const getExchangeJobsData = createSelector(
  getExchangeJobsList,
  getExchangeJobsTotal,
  (data, total) => ({ data, total })
);

export const getExchangeJobsResultsCount = createSelector(
  getExchangeJobsList,
  (data) => data.length
);

// Exchange Jobs Selectors, job family filter
export const getExchangeJobsFamilyFilterLoading = createSelector(
  getExchangeJobsFeature,
  fromExchangeJobsReducer.getJobFamilyFilterLoading
);

export const getExchangeJobsFamilyFilterLoadingSuccess = createSelector(
  getExchangeJobsFeature,
  fromExchangeJobsReducer.getJobFamilyFilterLoadingSuccess
);

export const getExchangeJobsFamilyFilterLoadingError = createSelector(
  getExchangeJobsFeature,
  fromExchangeJobsReducer.getJobFamilyFilterLoadingError
);

export const getExchangeJobFamilyFilterIsExpanded = createSelector(
  getExchangeJobsFeature,
  fromExchangeJobsReducer.getJobFamilyFilterIsExpanded
);

export const getExchangeJobFamilyFilterOptions = createSelector(
  getExchangeJobsFeature,
  fromExchangeJobsReducer.getJobFamilyFilterOptions
);

export const getExchangeJobFamilyFilterSelectedOptionNames = createSelector(
  getExchangeJobsFeature,
  (state) => state.jobFamilyOptions.filter(o => o.IsSelected).map(o => o.DisplayName)
);

// Company Jobs Selectors
export const getCompanyJobsGrid = createSelector(
  selectCompanyJobsState,
  (state: IFeatureGridState<fromCompanyJobsReducer.State>) => state.grid);

export const getCompanyJobsGridState = createSelector(
  getCompanyJobsGrid,
  (fromGridReducer.getGridState)
);

export const getCompanyJobsFeature = createSelector(
  selectCompanyJobsState,
  (state: IFeatureGridState<fromCompanyJobsReducer.State>) => state.feature);

export const {
  selectAll: getCompanyJobsList
} = fromCompanyJobsReducer.adapter.getSelectors(getCompanyJobsFeature);

export const getCompanyJobsTotal = createSelector(
  getCompanyJobsFeature,
  fromCompanyJobsReducer.getTotal);

export const getCompanyJobsData = createSelector(
  getCompanyJobsList,
  getCompanyJobsTotal,
  (data, total) => ({ data, total})
);

export const getCompanyJobsLoading = createSelector(
  getCompanyJobsFeature,
  (feature) => feature.loading
);

export const getCompanyJobsLoadingError = createSelector(
  getCompanyJobsFeature,
  (feature) => feature.loadingError
);

export const getCompanyJobsSearchTerm = createSelector(
  getCompanyJobsFeature,
  (feature) => feature.searchTerm
);

// detail panel
export const getCompanyJobsSelectedCompanyJobInDetailPanel = createSelector(
  getCompanyJobsFeature,
  (feature) => feature.selectedCompanyJobInDetailPanel
);

export const getCompanyJobsIsDetailPanelExpanded = createSelector(
  getCompanyJobsFeature,
  (feature) => feature.isDetailPanelExpanded
);

// associations
export const getSelectedCompanyJobs = createSelector(
    getCompanyJobsFeature,
    fromCompanyJobsReducer.getSelectedCompanyJobs);

export const getExchangeJobAssociations = createSelector(
    getExchangeJobsFeature,
    (feature) => feature.ExchangeJobAssociations
);
