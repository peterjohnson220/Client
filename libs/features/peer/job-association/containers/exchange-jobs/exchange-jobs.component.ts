import { Component, OnInit, OnDestroy, ViewChild, HostListener } from '@angular/core';

import { select, Store } from '@ngrx/store';
import { Observable, Subscription } from 'rxjs';
import { GridDataResult, GridComponent, DataStateChangeEvent, PageChangeEvent } from '@progress/kendo-angular-grid';
import { State } from '@progress/kendo-data-query';
import { TooltipDirective } from '@progress/kendo-angular-tooltip';

import { InputDebounceComponent } from 'libs/forms/components';
import * as fromJobAssociationReducers from '../../reducers';
import { GridTypeEnum, GenericMenuItem } from 'libs/models/common';

import * as fromExchangeJobsReducer from '../../reducers';
import * as exchangeJobsActions from '../../actions/exchange-jobs.actions';
import * as fromGridActions from 'libs/core/actions/grid.actions';
import { CompanyJob, ExchangeJob, ExchangeJobAssociation } from '../../models';

@Component({
  selector: 'pf-peer-job-association-exchange-jobs',
  templateUrl: './exchange-jobs.component.html',
  styleUrls: ['./exchange-jobs.component.scss']
})
export class ExchangeJobsComponent implements OnInit, OnDestroy {
  @ViewChild(TooltipDirective) public tooltipDir: TooltipDirective;
  @ViewChild(InputDebounceComponent) public jobTitleSearchComponent: InputDebounceComponent;
  @ViewChild(GridComponent) public grid: GridComponent;

  // Observables
  exchangeJobs$: Observable<GridDataResult>;
  totalPeerExchangeJobs$: Observable<number>;
  gridState$: Observable<State>;
  loading$: Observable<boolean>;
  loadingError$: Observable<boolean>;
  selectedCompanyJobs$: Observable<CompanyJob[]>;
  exchangeJobAssociations$: Observable<ExchangeJobAssociation[]>;
  gridResultsCount$: Observable<number>;
  selectedExchangeJob$: Observable<ExchangeJob>;
  isDetailPanelExpanded$: Observable<boolean>;

  // Subscriptions
  searchTermSubscription: Subscription;
  selectedCompanyJobsSubscription: Subscription;
  exchangeJobAssociationsSubscription: Subscription;
  selectedExchangeJobSubscription: Subscription;

  // Properties
  maxAssociableThreshold: number;
  isListView: boolean;
  isJobFamilyFilterExpanded: boolean;
  searchTerm: string;
  selectedCompanyJobs: CompanyJob[];
  exchangeJobAssociations: ExchangeJobAssociation[];
  selectedExchangeJob: ExchangeJob;

  // Job Family Filter
  isJobFamilyFilterExpandedSubscription: Subscription;
  isJobFamilyFilterExpanded$: Observable<boolean>;
  isJobFamilyFilterLoading$: Observable<boolean>;
  jobFamilyFilterOptions$: Observable<GenericMenuItem[]>;
  selectedJobFamilyOptionNames$: Observable<string[]>;

  constructor(private store: Store<fromJobAssociationReducers.State>) {}

  ngOnInit() {
    // Set properties
    this.maxAssociableThreshold = 10;
    this.isListView = true;

    // Register Observables
    this.loading$ = this.store.pipe(select(fromJobAssociationReducers.getExchangeJobsLoading));
    this.loadingError$ = this.store.pipe(select(fromJobAssociationReducers.getExchangeJobsLoadingError));
    this.exchangeJobs$ = this.store.pipe(select(fromJobAssociationReducers.getExchangeJobsData));
    this.totalPeerExchangeJobs$ = this.store.pipe(select(fromJobAssociationReducers.getExchangeJobsTotal));
    this.gridState$ = this.store.pipe(select(fromJobAssociationReducers.getExchangeJobsGridState));
    this.selectedCompanyJobs$ = this.store.pipe(select(fromJobAssociationReducers.getSelectedCompanyJobs));
    this.exchangeJobAssociations$ = this.store.pipe(select(fromJobAssociationReducers.getExchangeJobAssociations));
    this.gridResultsCount$ = this.store.pipe(select(fromJobAssociationReducers.getExchangeJobsResultsCount));
    this.selectedExchangeJob$ = this.store.pipe(select(fromJobAssociationReducers.getExchangeJobsSelectedExchangeJob));
    this.isDetailPanelExpanded$ = this.store.pipe(select(fromJobAssociationReducers.getExchangeJobsIsDetailPanelExpanded));

    //  Register Observables, job family
    this.isJobFamilyFilterExpanded$ = this.store.pipe(select(fromExchangeJobsReducer.getExchangeJobFamilyFilterIsExpanded));
    this.isJobFamilyFilterLoading$ = this.store.pipe(select(fromExchangeJobsReducer.getExchangeJobsFamilyFilterLoading));
    this.jobFamilyFilterOptions$ = this.store.pipe(select(fromExchangeJobsReducer.getExchangeJobFamilyFilterOptions));

    // Register Subscriptions
    this.searchTermSubscription = this.store.pipe(select(fromJobAssociationReducers.getExchangeJobsSearchTerm))
        .subscribe((searchTerm) => this.searchTerm = searchTerm);

    this.selectedCompanyJobsSubscription = this.selectedCompanyJobs$
        .subscribe((selectedCompanyJobs) => this.selectedCompanyJobs = selectedCompanyJobs);

    this.exchangeJobAssociationsSubscription = this.exchangeJobAssociations$
        .subscribe((exchangeJobAssociations) => this.exchangeJobAssociations = exchangeJobAssociations);

    this.selectedExchangeJobSubscription = this.selectedExchangeJob$
      .subscribe((selectedExchangeJob) => this.selectedExchangeJob = selectedExchangeJob);

    // job family
    this.isJobFamilyFilterExpanded$ =
      this.store.pipe(select(fromExchangeJobsReducer.getExchangeJobFamilyFilterIsExpanded));

    this.isJobFamilyFilterLoading$ =
      this.store.pipe(select(fromExchangeJobsReducer.getExchangeJobsFamilyFilterLoading));

    this.jobFamilyFilterOptions$ = this.store.pipe(select(fromExchangeJobsReducer.getExchangeJobFamilyFilterOptions));

    this.selectedJobFamilyOptionNames$ =
      this.store.pipe(select(fromExchangeJobsReducer.getExchangeJobFamilyFilterSelectedOptionNames));

    this.isJobFamilyFilterExpandedSubscription = this.isJobFamilyFilterExpanded$.subscribe((isExpanded) => {
      this.isJobFamilyFilterExpanded = isExpanded;
    });
  }

  ngOnDestroy() {
    this.searchTermSubscription.unsubscribe();
    this.selectedCompanyJobsSubscription.unsubscribe();
    this.exchangeJobAssociationsSubscription.unsubscribe();
    this.selectedExchangeJobSubscription.unsubscribe();
    // job family
    this.isJobFamilyFilterExpandedSubscription.unsubscribe();
  }

  reload(resetSearchTerm = false): void {
    // if this is invoked from an empty search results grid reset the term, otherwise keep the term as is and reload
    if (resetSearchTerm) {
      this.jobTitleSearchComponent.clearValue();
      this.store.dispatch(new exchangeJobsActions.ClearSelectedJobFamilies());
    } else {
      this.store.dispatch(new exchangeJobsActions.LoadExchangeJobs());
    }
  }

  showExchangeTitleTooltip(e: any): void {
    if (e.target.offsetWidth < e.target.scrollWidth && e.target.id.includes('exchange-title-')) {
      this.tooltipDir.toggle(e.target);
    } else {
      this.tooltipDir.hide();
    }
  }

  // event handlers
  handleDataStateChange(state: DataStateChangeEvent): void {
    this.store.dispatch(new fromGridActions.UpdateGrid(GridTypeEnum.JobAssociationModalPeerExchangeJobs, state));
    this.store.dispatch(new exchangeJobsActions.LoadExchangeJobs());
  }

  handleDetailExpand(event: any): void {
    this.store.dispatch(new exchangeJobsActions.CloseDetailPanel());

    // determine how many results we have in the grid
    const gridData = this.grid.data as any;
    const totalRows = gridData.data.length;

    // collapse all rows that are not the newly expanded row so we only have one detail open at a time
    for (let i = 0; i < totalRows; i++) {
      if (i !== event.index) {
        this.grid.collapseRow(i);
      }
    }
  }

  closeJobFamilyFilter(): void {
    if (this.isJobFamilyFilterExpanded) {
      this.store.dispatch(new exchangeJobsActions.ToggleJobFamilyFilter(false));
    }
  }

  handleJobFamilyToggle(isExpanded?: boolean) {
    this.store.dispatch(new exchangeJobsActions.ToggleJobFamilyFilter(isExpanded));
  }

  handleJobFamilyCheckboxToggle(option: GenericMenuItem) {
    this.store.dispatch(new exchangeJobsActions.ToggleJobFamilyFilterSelection(option));
    this.store.dispatch(new fromGridActions.PageChange(GridTypeEnum.JobAssociationModalPeerExchangeJobs, { skip: 0 } as PageChangeEvent));
    this.store.dispatch(new exchangeJobsActions.LoadExchangeJobs());
  }

  handleJobFamilyClearSelections() {
    this.store.dispatch(new exchangeJobsActions.ClearSelectedJobFamilies());
    this.store.dispatch(new fromGridActions.PageChange(GridTypeEnum.JobAssociationModalPeerExchangeJobs, { skip: 0 } as PageChangeEvent));
    this.store.dispatch(new exchangeJobsActions.LoadExchangeJobs());
  }

  handleExchangeJobClick(exchangeJob: ExchangeJob) {
    this.store.dispatch(new exchangeJobsActions.SelectExchangeJob(exchangeJob));
  }

  handleCloseDetailPanel() {
    this.store.dispatch(new exchangeJobsActions.CloseDetailPanel());
  }

  handleSearchFilterChanged(searchTerm: string): void {
    this.store.dispatch(new exchangeJobsActions.UpdateSearchTerm(searchTerm));

    // only search if 2+ chars are supplied, or if the term is reset and it's empty
    if (!searchTerm || searchTerm.length >= 2) {
      this.store.dispatch(new fromGridActions.PageChange(GridTypeEnum.JobAssociationModalPeerExchangeJobs, { skip: 0 } as PageChangeEvent));
      this.store.dispatch(new exchangeJobsActions.LoadExchangeJobs());
    }
  }

  handleAssociateClick(exchangeId: number, exchangeJobId: number): void {
    if (this.isAssociable(exchangeId, exchangeJobId)) {
      this.store.dispatch(new exchangeJobsActions.AddAssociation({
        ExchangeId: exchangeId,
        ExchangeJobId: exchangeJobId,
        CompanyJobs: this.selectedCompanyJobs
      }));
    }
  }

  handleRemoveAssociateClick(exchangeId: number, exchangeJobId: number, companyJobId: number): void {
    this.store.dispatch(new exchangeJobsActions.RemoveAssociation(
        exchangeId,
        exchangeJobId,
        companyJobId));
  }

  // close the job family filter when the escape key is clicked
  @HostListener('document:keyup', ['$event'])
  handleKeyboardEvent(event: KeyboardEvent) {
    if (event.key.toLowerCase() === 'escape') {
      this.store.dispatch(new exchangeJobsActions.ToggleJobFamilyFilter(false));
      this.store.dispatch(new exchangeJobsActions.CloseDetailPanel());
    }
  }

  // validation/business logic
  isAssociable(exchangeId: number, exchangeJobId: number): boolean {
    let isAssociable = false;
    if (this.hasSelectedCompanyJobs() &&
        this.selectedCompanyJobAreUniqueToExchange(exchangeId) &&
        this.selectedCompanyJobsAndAssociatedExchangeJobAreLessThanLimit(exchangeId, exchangeJobId) ) {
      isAssociable = true;
    }
    return isAssociable;
  }

  hasSelectedCompanyJobs(): boolean {
    return this.selectedCompanyJobs.length > 0;
  }

  selectedCompanyJobAreUniqueToExchange(exchangeId: number): boolean {
    const exchangeJobAssociations = this.exchangeJobAssociations.filter((x) => x.ExchangeId === exchangeId);
    const isUnique = true;

    if (exchangeJobAssociations.length === 0) {
      return isUnique;
    }

    const selectedCompanyJobIds = this.selectedCompanyJobs.map(x => x.CompanyJobId).sort();

    for (const exchangeJobAssociation of exchangeJobAssociations) {
      for (const companyJobId of exchangeJobAssociation.CompanyJobs.map(cj => cj.CompanyJobId).sort()) {
        if (selectedCompanyJobIds.indexOf(companyJobId) >= 0) {
          return !isUnique;
        }
      }
    }
    return isUnique;
  }

  selectedCompanyJobsAndAssociatedExchangeJobAreLessThanLimit(exchangeId: number, exchangeJobId: number): boolean {
    let count = this.selectedCompanyJobs.length;
    count += this.getAssociationCount(exchangeId, exchangeJobId);
    return count <= this.maxAssociableThreshold;
  }

  getAssociationCount(exchangeId: number, exchangeJobId: number): number {
    const exchangeJobAssociation = this.exchangeJobAssociations
        .filter((x) => x.ExchangeId === exchangeId && x.ExchangeJobId === exchangeJobId);

    if (exchangeJobAssociation.length <= 0) {
      return 0;
    }

    let totalAssociatedJobs = 0;
    for (let i = 0; i < exchangeJobAssociation.length; i++) {
      totalAssociatedJobs += exchangeJobAssociation[i].CompanyJobs.map((cj) => cj.CompanyJobId).length;
    }
    return totalAssociatedJobs;
  }

  getCompanyJobAssociations(exchangeId: number, exchangeJobId: number): CompanyJob[] {
    const companyJobs: CompanyJob[] = [];
    const exchangeJobAssociation = this.exchangeJobAssociations
        .filter((x) => x.ExchangeId === exchangeId && x.ExchangeJobId === exchangeJobId);

    if (exchangeJobAssociation.length <= 0) {
      return companyJobs;
    }

    for (let i = 0; i < exchangeJobAssociation.length; i++) {
      exchangeJobAssociation[i].CompanyJobs.forEach((cj) => companyJobs.push(cj));
    }
    return companyJobs;
  }

  toggleViewType(isListView: boolean): void {
    if (this.isListView === isListView) {
      return;
    }
    this.isListView = !this.isListView;
  }
}
