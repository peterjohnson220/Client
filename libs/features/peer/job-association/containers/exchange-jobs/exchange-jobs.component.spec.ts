import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';

import { combineReducers, Store, StoreModule } from '@ngrx/store';
import { of } from 'rxjs';
import { DataStateChangeEvent } from '@progress/kendo-angular-grid';
import { NgbTooltipModule } from '@ng-bootstrap/ng-bootstrap';
import { GridComponent } from '@progress/kendo-angular-grid';

import { HighlightTextPipe, JobDescriptionParserPipe } from 'libs/core/pipes';
import * as fromRootState from 'libs/state/state';
import * as fromGridActions from 'libs/core/actions/grid.actions';
import { GridTypeEnum } from 'libs/models/common';

import * as fromExchangeJobsActions from '../../actions/exchange-jobs.actions';
import * as fromExchangeJobsReducer from '../../reducers';
import { ExchangeJobsComponent } from './exchange-jobs.component';
import { generateMockCompanyJob } from '../../models';

describe('ExchangeJobsComponent', () => {
  let component: ExchangeJobsComponent;
  let fixture: ComponentFixture<ExchangeJobsComponent>;
  let store: Store<fromExchangeJobsReducer.State>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        StoreModule.forRoot({
          ...fromRootState.reducers,
          feature_job_association: combineReducers(fromExchangeJobsReducer.reducers)
        }),
        NgbTooltipModule
      ],
      declarations: [ ExchangeJobsComponent, HighlightTextPipe, JobDescriptionParserPipe ],
      schemas: [ NO_ERRORS_SCHEMA ]
    })
    .compileComponents();

    store = TestBed.get(Store);
    spyOn(store, 'dispatch');
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ExchangeJobsComponent);
    component = fixture.componentInstance;
    component.jobTitleSearchComponent = {} as any;
    component.jobTitleSearchComponent.writeValue = () => {};
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should show total exchange jobs', () => {
    component.totalPeerExchangeJobs$ = of(123);

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should fire the right actions when the grid state is changed', () => {
    const dataStateChangeEvent = {} as DataStateChangeEvent;
    component.handleDataStateChange(dataStateChangeEvent);

    const updateGridAction = new fromGridActions.UpdateGrid(GridTypeEnum.JobAssociationModalPeerExchangeJobs, dataStateChangeEvent);
    const loadExchangeJobsAction = new fromExchangeJobsActions.LoadExchangeJobs();

    expect(store.dispatch).toHaveBeenCalledWith(updateGridAction);
    expect(store.dispatch).toHaveBeenCalledWith(loadExchangeJobsAction);
  });

  it('should fire the UpdateSearchTerm action when the search term is changed', () => {
    const searchTerm = 'abc';
    component.handleJobTitleFilterChanged(searchTerm);

    const updateSearchTermAction = new fromExchangeJobsActions.UpdateSearchTerm(searchTerm);

    expect(store.dispatch).toHaveBeenCalledWith(updateSearchTermAction);
  });

  it('should not fire the LoadExchangeJobs action when the search term is one character', () => {
    const searchTerm = 'a';
    component.handleJobTitleFilterChanged(searchTerm);

    const loadExchangeJobsAction = new fromExchangeJobsActions.LoadExchangeJobs();

    expect(store.dispatch).not.toHaveBeenCalledWith(loadExchangeJobsAction);
  });

  it('should fire the correct actions when the search term is more than one character', () => {
    const searchTerm = 'ab';
    component.handleJobTitleFilterChanged(searchTerm);

    const updateSearchTermAction = new fromExchangeJobsActions.UpdateSearchTerm(searchTerm);
    const loadExchangeJobsAction = new fromExchangeJobsActions.LoadExchangeJobs();

    expect(store.dispatch).toHaveBeenCalledWith(updateSearchTermAction);
    expect(store.dispatch).toHaveBeenCalledWith(loadExchangeJobsAction);
  });

  it('should fire the expected action when handleAssociateClick is clicked', () => {
    const dataItem = { ExchangeId: 123, ExchangeJobId: 456, CompanyJobMappings: [generateMockCompanyJob()] };

    const addAssociationAction = new fromExchangeJobsActions.AddAssociation({
      ExchangeId: dataItem.ExchangeId,
      ExchangeJobId: dataItem.ExchangeJobId,
      CompanyJobs: [generateMockCompanyJob()]
    });

    component.isAssociable = () => true;
    component.grid = {} as GridComponent;
    component.grid.expandRow = () => ({});
    component.grid.collapseRow = () => ({});
    component.selectedCompanyJobs = [generateMockCompanyJob()];
    component.handleAssociateClick(dataItem, 44);

    expect(store.dispatch).toHaveBeenCalledWith(addAssociationAction);
  });

  it('should not fire an action when handleAssociateClick is clicked and isAssociable returns false', () => {
    const dataItem = { ExchangeId: 123, ExchangeJobId: 456, CompanyJobMappings: [generateMockCompanyJob()] };

    const addAssociationAction = new fromExchangeJobsActions.AddAssociation({
      ExchangeId: dataItem.ExchangeId,
      ExchangeJobId: dataItem.ExchangeJobId,
      CompanyJobs: [generateMockCompanyJob()]
    });

    component.isAssociable = () => false;
    component.handleAssociateClick(dataItem, 44);

    expect(store.dispatch).not.toHaveBeenCalledWith(addAssociationAction);
  });

  it('should create the correct associate button tooltip msg when no jobs are selected', () => {
    component.selectedCompanyJobs = [];

    const tooltipText = component.createAssociateButtonTooltipText(123, 456);
    expect(tooltipText).toBe('First select the company job you want to associate');
  });

  it('should create the correct associate button tooltip msg when the job is associable', () => {
    component.maxAssociableThreshold = 10;
    component.selectedCompanyJobs = [generateMockCompanyJob()];
    component.isAssociable = () => true;
    component.getAssociationCount = () => 1;

    const tooltipText = component.createAssociateButtonTooltipText(123, 456);
    expect(tooltipText).toBe('Click to associate');
  });

  it('should create the correct associate button tooltip msg when there are > 10 associations', () => {
    component.maxAssociableThreshold = 10;
    component.selectedCompanyJobs = [generateMockCompanyJob()];
    component.isAssociable = () => false;
    component.getAssociationCount = () => component.maxAssociableThreshold + 1;

    const tooltipText = component.createAssociateButtonTooltipText(123, 456);
    expect(tooltipText).toBe('Exchange jobs should not have more than 10 associations per exchange');
  });

  it('should create the correct associate button tooltip msg when the job is already associated to an exchange', () => {
    component.maxAssociableThreshold = 10;
    component.selectedCompanyJobs = [generateMockCompanyJob()];
    component.isAssociable = () => false;
    component.getAssociationCount = () => 1;

    const tooltipText = component.createAssociateButtonTooltipText(123, 456);
    expect(tooltipText).toBe('A single company job can be associated to only 1 job per exchange');
  });
});
