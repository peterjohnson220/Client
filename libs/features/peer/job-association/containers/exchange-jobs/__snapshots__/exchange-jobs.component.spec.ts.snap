// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeJobsComponent should show total exchange jobs 1`] = `
<pf-peer-job-association-exchange-jobs
  exchangeJobAssociations={[Function Array]}
  exchangeJobAssociations$={[Function Store]}
  exchangeJobAssociationsSubscription={[Function Subscriber]}
  exchangeJobs$={[Function Store]}
  grid="undefined"
  gridResultsCount$={[Function Store]}
  gridState$={[Function Store]}
  isDetailPanelExpanded$={[Function Store]}
  isJobFamilyFilterExpanded="false"
  isJobFamilyFilterExpanded$={[Function Store]}
  isJobFamilyFilterExpandedSubscription={[Function Subscriber]}
  isJobFamilyFilterLoading$={[Function Store]}
  isListView={[Function Boolean]}
  jobFamilyFilterOptions$={[Function Store]}
  jobTitleSearchComponent="undefined"
  loading$={[Function Store]}
  loadingError$={[Function Store]}
  loadingPreviousAssociations$={[Function Store]}
  loadingPreviousAssociationsError$={[Function Store]}
  loadingPreviousAssociationsSuccess$={[Function Store]}
  maxAssociableThreshold={[Function Number]}
  previousAssociations$={[Function Store]}
  searchTerm=""
  searchTermSubscription={[Function Subscriber]}
  selectedCompanyJobs={[Function Array]}
  selectedCompanyJobs$={[Function Store]}
  selectedCompanyJobsSubscription={[Function Subscriber]}
  selectedExchangeJob={[Function Object]}
  selectedExchangeJob$={[Function Store]}
  selectedExchangeJobSubscription={[Function Subscriber]}
  selectedJobFamilyOptionNames$={[Function Store]}
  store={[Function Store]}
  tooltipDir="undefined"
  totalPeerExchangeJobs$={[Function Store]}
>
  <h2
    class="text-center"
  >
    Peer Exchange Jobs
  </h2><div
    class="filters position-relative d-flex mt-2 mb-2"
  >
    <div>
      <i
        class="fas fa-search position-absolute"
      />
      <pf-input-debounce
        id="exchange-jobs-search-box"
      />
    </div>
    <div
      class="ml-3"
    >
      <pf-multi-select
        pfclickelsewhere=""
      />
    </div>
  </div><div>
    <label
      class="text-uppercase"
      data-qa-id="lbl-exchange-jobs-total"
    >
       peer exchange jobs: 0 total 
    </label>
    <button
      class="view-button ml-2 invisible view-button-active"
      ng-reflect-klass="view-button ml-2 invisible"
      ng-reflect-ng-class="[object Object]"
      title="Grid View"
    >
      <i
        class="fas fa-table"
      />
    </button>
    <button
      class="view-button invisible"
      ng-reflect-klass="view-button invisible"
      ng-reflect-ng-class="[object Object]"
      title="Detail View"
    >
      <i
        class="fas fa-list"
      />
    </button>
  </div><div
    class="grid-container position-relative"
  >
    <pf-async-container
      class="h-100"
    >
      
      <div
        class="k-tooltip-container"
        filter=".k-grid td"
        kendotooltip=""
        showon="none"
      >
        <pf-grid-detail-panel />
        <kendo-grid
          style="height: 100%;"
        >
          <kendo-grid-column>
            
          </kendo-grid-column>
          <kendo-grid-column
            field="ExchangeJobTitle"
            title="Job Title"
          >
            
          </kendo-grid-column>
          <kendo-grid-column
            field="ExchangeJobFamily"
            title="Job Family"
          >
            
          </kendo-grid-column>
          <kendo-grid-column
            field="ExchangeName"
            title="Exchange Name"
          />
          
          
        </kendo-grid>
      </div>
      
    </pf-async-container>
  </div>
</pf-peer-job-association-exchange-jobs>
`;
